primitives/src/traits.rs

```diff
@@ -99,10 +126,34 @@ pub trait AMM<AccountId, AssetId, AssetPair, Amount> {
 }
 
 pub trait Resolver<AccountId, Intention, E> {
-	/// Resolve an intention directl via AMM pool.
+	/// Resolve an intention directly via AMM pool.
 	fn resolve_single_intention(intention: &Intention);
 
 	/// Resolve intentions by either directly trading with each other or via AMM pool.
 	/// Intention ```intention``` must be validated prior to call this function.
 	fn resolve_matched_intentions(pair_account: &AccountId, intention: &Intention, matched: &[Intention]);
 }
+
+/// AMM handlers used by AMM pools to indicate various events.
+pub trait AMMHandlers<AccountId, AssetId, AssetPair, Balance> {
+	/// Register an asset to be handled by price-oracle pallet.
+	/// If an asset is not registered, calling `on_trade` results in populating the price buffer in the price oracle pallet,
+	/// but the entries are ignored and the average price for the asset is not calculated.
+	fn on_create_pool(asset_pair: AssetPair);
+	/// Include a trade in the average price calculation of the price-oracle pallet.
+	fn on_trade(amm_transfer: &AMMTransfer<AccountId, AssetId, AssetPair, Balance>, liq_amount: Balance);
+	/// Known `on_trade` overhead. Needs to be specified here if we don't want to make AMM pools tightly coupled with the price oracle pallet.
+	fn on_trade_weight() -> Weight;
+}
+
+impl<AccountId, AssetId, AssetPair, Balance> AMMHandlers<AccountId, AssetId, AssetPair, Balance> for () {
+	fn on_create_pool(_asset_pair: AssetPair) {}
+	fn on_trade(_amm_transfer: &AMMTransfer<AccountId, AssetId, AssetPair, Balance>, _liq_amount: Balance) {}
+	fn on_trade_weight() -> Weight {
```

Comment:
Better descriptions here.. I still can't tell when the `on_trade_weight` is actually called and what is supposed to do.